From 93f392563a75c2ced1503f625b3b885ac7800b6d Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Miroslav=20S=CC=8Cafa=CC=81r=CC=8C?=
 <miroslav.safar@znojmo.cz>
Date: Wed, 22 Nov 2023 15:36:09 +0100
Subject: [PATCH] feat: 1.19.4 player session id fix


diff --git a/protocol/src/main/java/net/md_5/bungee/protocol/Protocol.java b/protocol/src/main/java/net/md_5/bungee/protocol/Protocol.java
index 22769c88..21e156bf 100644
--- a/protocol/src/main/java/net/md_5/bungee/protocol/Protocol.java
+++ b/protocol/src/main/java/net/md_5/bungee/protocol/Protocol.java
@@ -511,7 +511,7 @@ public enum Protocol
                     PlayerSession.class,
                     PlayerSession::new,
                     map( ProtocolConstants.MINECRAFT_1_19_3, 0x20 ),
-                    map( ProtocolConstants.MINECRAFT_1_20, 0x06 )
+                    map( ProtocolConstants.MINECRAFT_1_19_4, 0x06 )
             );
             // QPlay end
 
-- 
2.39.2 (Apple Git-143)

